### PR TITLE
feature: Make email optional

### LINKED ISSUE #66: Cannot set user without email
Most of our users use twitter to login. Twitter does not return emails so we have had it optional in our past setup.

Currently switching to loopback and running into this modul erring when userObj.email is null (which is our default, a lot of our system depends on this so it is not possible to change it).

Would a PR be accepted to add a flag to make email optional?
[here](https://github.com/strongloop/loopback-component-passport/blob/master/lib/models/user-identity.js#L116) and [here](https://github.com/strongloop/loopback-component-passport/blob/master/lib/models/user-identity.js#L50)

from:
```js
// login func
if (!userObj.email) {
    return cb('email is missing from the user profile');
}

// profileToUser
if (options.emailOptional && !email) {
      // Fake an e-mail
      email = (profile.username || profile.id) + '@loopback.' +
              (profile.provider || provider) + '.com';
}
```
to

```js
// login func
if (!userObj.email && !options.emailOptional) {
    return cb('email is missing from the user profile');
}

// profileToUser
if (!email && !options.emailOptional) {
      // Fake an e-mail
      email = (profile.username || profile.id) + '@loopback.' +
              (profile.provider || provider) + '.com';
}
```

cc @terakilobyte @QuincyLarson

### PR DESCRIPTION
**Pull request by BerkeleyTrue**
Only add email field to user object if available and not optional. This way emails can still be indexed with sparse indexing.

closes #66

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by raymondfeng** [comment]
@BerkeleyTrue Please sign the CLA at - https://cla.strongloop.com/agreements/strongloop/loopback-component-passport

**Comment by BerkeleyTrue** [comment]
signed

File: lib/models/user-identity.js [hunk 1]
```diff
@@ -112,8 +114,8 @@
       var userModel = (userIdentityModel.relations.user &&
                        userIdentityModel.relations.user.modelTo) ||
                        loopback.getModelByType(loopback.User);
-      var userObj = (options.profileToUser || profileToUser)(provider, profile);
-      if (!userObj.email) {
+      var userObj = (options.profileToUser || profileToUser)(provider, profile, options);
+      if (!userObj.email && !options.emailOptional) {
         return cb('email is missing from the user profile');
```
Comment (raymondfeng) [inline lib/models/user-identity.js]: Let's do this shortcut within a `process.nextTick` block.

Comment (BerkeleyTrue) [inline lib/models/user-identity.js]: Not sure what you mean. wrap line 117 in a process.nextTick?

**Comment by BerkeleyTrue** [comment]
Had to change the query structure. In certain situations user identities where being assigned to random user documents. I also wrapped the callback in a nextTick

**Comment by pongstr** [comment]
:+1:

**Comment by raymondfeng** [comment]
ok to test

**Comment by BerkeleyTrue** [comment]
@raymondfeng Any feedback on this?

**Comment by BerkeleyTrue** [comment]
woot!

File: lib/models/user-identity.js [hunk 2]
```diff
@@ -112,14 +114,23 @@
       var userModel = (userIdentityModel.relations.user &&
                        userIdentityModel.relations.user.modelTo) ||
                        loopback.getModelByType(loopback.User);
-      var userObj = (options.profileToUser || profileToUser)(provider, profile);
-      if (!userObj.email) {
-        return cb('email is missing from the user profile');
+      var userObj = (options.profileToUser || profileToUser)(provider, profile, options);
+      if (!userObj.email && !options.emailOptional) {
+        process.nextTick(function() {
```
Comment (alexpls) [inline lib/models/user-identity.js]: We're integrating loopback-component-passport into our Loopback app and are having some trouble understanding this code. Why is the `nextTick` required here - it seems like it allows the execution to fall through and to create the member even if the email is not available.

Comment (BerkeleyTrue) [inline lib/models/user-identity.js]: @alexpit You are correct, this is indeed a bug. There is a missing return before the process.nextTick.

Can you file an issue and maybe create a PR?

Comment (alexpls) [inline lib/models/user-identity.js]: @BerkeleyTrue will do - thanks very much for your response!

Comment (BerkeleyTrue) [inline lib/models/user-identity.js]: Thanks. I'm surprised this hasn't been caught earlier.

**Pull request merged**